tools/funccount.py

```diff
@@ -57,18 +57,25 @@ def signal_ignore(signal, frame):
 bpf_text = """
 #include <uapi/linux/ptrace.h>
 
-struct key_t {
+#define MAP_SIZE 10000
+
+struct val_t {
     u64 ip;
+    u64 value;
 };
-BPF_HASH(counts, struct key_t);
+BPF_TABLE("array", int, struct val_t, counts, MAP_SIZE);
 
 int trace_count(struct pt_regs *ctx) {
     FILTER
-    struct key_t key = {};
-    u64 zero = 0, *val;
-    key.ip = ctx->ip;
-    val = counts.lookup_or_init(&key, &zero);
-    (*val)++;
+    struct val_t *val;
+    u64 zero = 0;
+    u32 key = ctx->ip % MAP_SIZE;
```

Comment:
Or have a read-only (from kernel point of view) map that gives pc to idx
translation.
On Feb 23, 2016 10:41 AM, "4ast" <notifications@github.com> wrote:

> In tools/funccount.py
> <https://github.com/iovisor/bcc/pull/415#discussion_r53826940>:
>
> >
> >  int trace_count(struct pt_regs *ctx) {
> >      FILTER
> > -    struct key_t key = {};
> > -    u64 zero = 0, *val;
> > -    key.ip = ctx->ip;
> > -    val = counts.lookup_or_init(&key, &zero);
> > -    (*val)++;
> > +    struct val_t *val;
> > +    u64 zero = 0;
> > +    u32 key = ctx->ip % MAP_SIZE;
>
> what do you guys mean by pre-indexing? have different program with hard
> coded index for every counted function?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/iovisor/bcc/pull/415/files#r53826940>.
>